x/archivedb/db.go

```diff
@@ -0,0 +1,150 @@
+//			Delete(foo)
+//
+// When requesting `Get(foo, 9)` it will return an errNotFound error because foo
+// was not defined at block height 9, it was defined later. When calling
+// `Get(foo, 99)` it will return a tuple `("foo's value is bar", 10)` returning
+// the value of `foo` at height 99 (which was set at height 10). If requesting
+// `Get(foo, 2000)` it will return an error because `foo` was deleted at height
+// 1000.
+type archiveDB struct {
+	// Must be held when reading/writing fields.
+	lock sync.RWMutex
+
+	ctx context.Context
+
+	rawDB database.Database
+}
+
+type batchWithHeight struct {
+	db     *archiveDB
+	height uint64
+	batch  database.Batch
+}
+
+func newDatabase(
+	ctx context.Context,
+	db database.Database,
+) (*archiveDB, error) {
+	return &archiveDB{
+		ctx:   ctx,
+		rawDB: db,
+	}, nil
+}
+
+// Fetches the value of a given prefix at a given height.
+//
+// If the value does not exists or it was actually removed an error is returned.
+// Otherwise a value does exists it will be returned, alongside with the height
+// at which it was updated prior the requested height.
+func (db *archiveDB) Get(key []byte, height uint64) ([]byte, uint64, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	internalKey, err := NewKey(key, height)
+	if err != nil {
+		return nil, 0, err
+	}
+	iterator := db.rawDB.NewIteratorWithStart(internalKey.Bytes)
+	if !iterator.Next() {
+		// There is no available key with the requested prefix
+		return nil, 0, database.ErrNotFound
+	}
+
+	internalKey, err = ParseKey(iterator.Key())
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if !bytes.Equal(internalKey.Prefix, key) || internalKey.IsDeleted {
+		// The current key has either a different prefix or the found key has a
+		// deleted flag.
+		//
+		// The previous key that was found does has another prefix, because the
+		// iterator is not aware of prefixes. If this happens it means the
+		// prefix at the requested height does not exists.
+		//
+		// The database is append only, so when removing a record creates a new
+		// record with an special flag is being created. Before returning the
+		// value we check if the deleted flag is present or not.
+		return nil, 0, database.ErrNotFound
+	}
+
+	return iterator.Value(), internalKey.Height, nil
+}
+
+// Creates a new batch to append database changes in a given height
+func (db *archiveDB) NewBatch(height uint64) batchWithHeight {
+	return batchWithHeight{
+		db:     db,
+		height: height,
+		batch:  db.rawDB.NewBatch(),
+	}
+}
+
+// Writes the changes to the database
+func (c *batchWithHeight) Write() error {
+	c.db.lock.Lock()
+	defer c.db.lock.Unlock()
+	return c.batch.Write()
+}
+
+// Delete any previous state that may be stored in the database
+func (c *batchWithHeight) Delete(key []byte) error {
+	internalKey, err := NewKey(key, c.height)
+	if err != nil {
+		return err
+	}
+	if err = internalKey.SetDeleted(); err != nil {
+		return err
+	}
+	return c.batch.Put(internalKey.Bytes, []byte{})
```

Comment:
we should also follow the semantics for whether slices can be modified after they are passed to the batch per the db interface.

Reply:
The `Batch` that is returned by this database is a thin wrapper on top the existing `Batch` that is available everywhere in the code. It has to extend it to keep track of the current height. The key provider to the `Batch` is also modified and passed over to the low level `Batch`.  The modifications looks as follow

```
<key: var length><MaxUint64 - height:8 bytes><isDeleted: 1 byte>
```

The key has this extra information attached to itself to make it unique (`foo` at height 1 is different than `foo` at height 3). Aside from making it different, because they are different entries, it will also leverages how the database stores things internally, adding a natural sorting. `foo:3` will come before to `foo:1`. 

Is this explanation good enough?

Reply:
> how is the behavior handled if we do eg,
> 
> ```
> Put(key, v1)
> Delete(key)
> Put(key, v2)
> ```
> 
> ? I think we should follow the behavior of the database's batch interface here.

The interface will write the Delete() and Last put() but when reading only the delete will be read. I think the Batch should prevent to queue another multiple operations per prefixes and error when that happens. Thoughts @darioush @aaronbuchwald ?